eth/handler_bor.go

```
@@ -39,80 +39,148 @@ var (
 	//nolint:prealloc
 	var (
-		blockNums   []uint64      = make([]uint64, 0)
-		blockHashes []common.Hash = make([]common.Hash, 0)
+		blockNum  uint64
+		blockHash common.Hash
 	)
 
 	// Fetch the checkpoint count from heimdall
 	count, err := bor.HeimdallClient.FetchCheckpointCount(ctx)
 	if err != nil {
 		log.Debug("Failed to fetch checkpoint count for whitelisting", "err", err)
-		return blockNums, blockHashes, errCheckpointCount
+		return blockNum, blockHash, errCheckpointCount
+	}
+
+	if count == 0 {
+		return blockNum, blockHash, errNoCheckpoint
+	}
+
+	// fetch the latest checkpoint from Heimdall
+	checkpoint, err := bor.HeimdallClient.FetchCheckpoint(ctx)
+	if err != nil {
+		log.Debug("Failed to fetch latest checkpoint for whitelisting", "err", err)
+		return blockNum, blockHash, errMilestone
+	}
+
+	// check if we have the checkpoint blocks
+	head := h.ethAPI.BlockNumber()
+	if head < hexutil.Uint64(checkpoint.EndBlock.Uint64()) {
+		log.Debug("Head block behind checkpoint block", "head", head, "milestone end block", checkpoint.EndBlock)
+		return blockNum, blockHash, errMissingMilestone
+	}
+
+	// verify the root hash of checkpoint
+	roothash, err := h.ethAPI.GetRootHash(ctx, checkpoint.StartBlock.Uint64(), checkpoint.EndBlock.Uint64())
+	if err != nil {
+		log.Debug("Failed to get root hash of checkpoint while whitelisting", "err", err)
+		return blockNum, blockHash, errRootHash
+	}
+
+	if roothash != checkpoint.RootHash.String()[2:] {
+		log.Warn("Checkpoint root hash mismatch while whitelisting", "expected", checkpoint.RootHash.String()[2:], "got", roothash)
+		return blockNum, blockHash, errCheckpointRootHashMismatch
+	}
+
+	// fetch the end checkpoint block hash
+	block, err := h.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(checkpoint.EndBlock.Uint64()), false)
+	if err != nil {
+		log.Debug("Failed to get end block hash of checkpoint while whitelisting", "err", err)
+		return blockNum, blockHash, errEndBlock
```

why don't what an error and remove logging on that level?

---

Do you mean we should return the err itself rather than creating errEndBlock

---

Like add the context to the error with fmt Errorf + %w